libs/ts-rest/core/src/lib/client.ts

```diff
@@ -133,23 +169,52 @@ const createFormData = (body: unknown) => {
   return formData;
 };
 
-export const fetchApi = (
-  path: string,
-  clientArgs: ClientArgs,
-  route: AppRoute,
-  body: unknown
-) => {
-  const apiFetcher = clientArgs.api || defaultApi;
+export const fetchApi = ({
+  path,
+  clientArgs,
+  route,
+  body,
+  extraInputArgs,
+  headers,
+}: {
+  path: string;
+  clientArgs: ClientArgs;
+  route: AppRoute;
+  body: unknown;
+  extraInputArgs: Record<string, unknown>;
+  headers: Record<string, string | undefined>;
+}) => {
+  const apiFetcher = clientArgs.api || tsRestFetchApi;
+
+  /**
+   * Combined headers should combine baseHeaders and headers from the api call, but
+   * if there are any duplicate keys, the headers from the api call should take precedence
+   * over the baseHeaders
+   *
+   * if the header is undefined, it should unset the header
+   */
+  const combinedHeaders = new Map();
+
+  Object.entries(clientArgs.baseHeaders).forEach(([key, value]) => {
+    combinedHeaders.set(key, value);
+  });
+
+  Object.entries(headers).forEach(([key, value]) => {
+    if (value === undefined) {
+      combinedHeaders.delete(key);
+    } else {
+      combinedHeaders.set(key, value);
+    }
+  });
```

Comment:
Can we simplify this to something like this?

```suggestion
  const combinedHeaders = {
    ...clientArgs.baseHeaders,
    ...headers
  };
```

I'm not 100% sure how `fetch` would handle undefined header values, but I'd take a very strong guess that it simply omits the header.

Also we'd probably need to normalize the header names to lower case in either implementations, otherwise we can possibly set the same header twice without `headers` overwriting a header if the casing was different. I'd also guess if this happens, the second differently-cased header would overwrite the first internally, but I'd rather not depend on undefined behavior of `fetch`

Reply:
Does header casing matter on the server side, though? I can't remember if frameworks like express and Nest auto-lowercase incoming headers?

Reply:
@michaelangrivera it doesn't but this is mainly about JS object keys being case sensitive, in this case we can have weird behavior if the same header is passed twice but with different casing. To combat this, a lot of JS client-side libraries, including fetch, normalize all headers to lower case.

@oliverbutler I just tested it, and it seems that a repeated header with different name casing will append the values together.

```typescript
// baseHeaders = { 'content-type': 'application/json' }
// headers     = { 'Content-Type': 'text/plain' }
const combinedHeaders = {
  ...clientArgs.baseHeaders,
  ...headers
};
```

This will result in the header being set as `content-type: application/json, text/plain`